talismane_core/src/main/java/com/joliciel/talismane/tokeniser/TokenSequence.java

```diff
@@ -28,54 +29,97 @@
 /**
  * A sequence of tokens. Note: by default, List iteration and associated methods
  * will only return non-whitespace tokens. For a list that includes whitespace
- * tokens, use the listWithWhiteSpace() method.
+ * tokens, use the listWithWhiteSpace() method.<br/>
+ * <br/>
+ * Only a single attribute of a given key can be added per token. If two
+ * attribute placeholders overlap and assign the same key, the one with the
+ * lower start index is privileged for all tokens, and the other one is skipped
+ * for all tokens. If they have the same start index, the one with the higher
+ * end index is priveleged for all tokens, and the other one is skipped for all
+ * tokens.
  * 
  * @author Assaf Urieli
  *
  */
-public class TokenSequence extends ArrayList<Token> implements Serializable {
+public class TokenSequence extends ArrayList<Token>implements Serializable {
 	private static final Logger LOG = LoggerFactory.getLogger(TokenSequence.class);
 	private static final long serialVersionUID = 1L;
 
 	private List<Integer> tokenSplits;
-	private String text;
-	private Sentence sentence;
-	private List<Token> listWithWhiteSpace = new ArrayList<Token>();
+	private final Sentence sentence;
+	private final List<Token> listWithWhiteSpace;
 	private List<Token> tokensAdded = null;
 	private double score = 0;
 	private boolean scoreCalculated = false;
-	TokenisedAtomicTokenSequence underlyingAtomicTokenSequence;
+	private TokenisedAtomicTokenSequence underlyingAtomicTokenSequence;
 	private Integer atomicTokenCount = null;
 	private boolean finalised = false;
 	private boolean sentenceTagsAdded = false;
 	private boolean withRoot = false;
 	private PosTagSequence posTagSequence;
-	protected boolean textProvided = false;
+	private final Map<Integer, Annotation<TokenPlaceholder>> placeholderMap;
+	@SuppressWarnings("rawtypes")
+	private final Map<String, NavigableSet<Annotation<TokenAttribute>>> attributeOrderingMap;
 
 	private PosTaggerLexicon lexicon;
 
 	private final TalismaneSession talismaneSession;
 
-	/**
-	 * Create a token sequence for a given sentence.
-	 */
-	public TokenSequence(Sentence sentence, TalismaneSession talismaneSession) {
+	TokenSequence(Sentence sentence, TalismaneSession talismaneSession) {
 		this.sentence = sentence;
 		this.talismaneSession = talismaneSession;
-		this.text = sentence.getText();
+		this.placeholderMap = new HashMap<>();
+		this.listWithWhiteSpace = new ArrayList<>();
+
+		List<Annotation<TokenPlaceholder>> placeholders = sentence.getAnnotations(TokenPlaceholder.class);
+		if (placeholders != null) {
+			for (Annotation<TokenPlaceholder> placeholder : placeholders) {
+				// take the first placeholder at this start index only
+				// thus declaration order is the order at which they're
+				// applied
+				if (!placeholderMap.containsKey(placeholder.getStart()))
+					placeholderMap.put(placeholder.getStart(), placeholder);
+			}
+		}
+
+		@SuppressWarnings("rawtypes")
```

Comment:
Is this `SuppressWarnings` needed? The `getAnnotations` seems to return a fully typed list.